worker/raft/rafttest/fixtures.go

```diff
@@ -33,39 +34,53 @@ func (s *RaftFixture) SetUpTest(c *gc.C) {
 	s.IsolationSuite.SetUpTest(c)
 	c.Assert(s.FSM, gc.NotNil, gc.Commentf("FSM must be set by embedding test suite"))
 
-	s.Store = raft.NewInmemStore()
-	s.SnapshotStore = raft.NewInmemSnapshotStore()
-	addr, transport := raft.NewInmemTransport("machine-0")
-	s.Transport = transport
-	s.AddCleanup(func(*gc.C) { s.Transport.Close() })
-
-	raftConfig := raft.DefaultConfig()
-	raftConfig.HeartbeatTimeout = 100 * time.Millisecond
-	raftConfig.ElectionTimeout = raftConfig.HeartbeatTimeout
-	raftConfig.LeaderLeaseTimeout = raftConfig.HeartbeatTimeout
-	raftConfig.LocalID = raft.ServerID(string(addr))
-	raftConfig.Logger = log.New(&raftutil.LoggoWriter{
-		loggo.GetLogger("juju.worker.raft.raftclusterer_test"),
-		loggo.DEBUG,
-	}, "", 0)
-
-	r, err := raft.NewRaft(raftConfig, s.FSM, s.Store, s.Store, s.SnapshotStore, s.Transport)
-	c.Assert(err, jc.ErrorIsNil)
-	s.AddCleanup(func(c *gc.C) {
-		f := r.Shutdown()
-		c.Assert(f.Error(), jc.ErrorIsNil)
-	})
-	s.Raft = r
-
+	s.Raft, s.Config, s.Transport, s.Store, s.SnapshotStore = s.NewRaft(c, "machine-0", s.FSM)
 	c.Assert(s.Raft.BootstrapCluster(raft.Configuration{
 		Servers: []raft.Server{{
-			ID:      raftConfig.LocalID,
-			Address: addr,
+			ID:      s.Config.LocalID,
+			Address: s.Transport.LocalAddr(),
 		}},
 	}).Error(), jc.ErrorIsNil)
 	select {
-	case <-s.Raft.LeaderCh():
+	case leader := <-s.Raft.LeaderCh():
+		c.Assert(leader, jc.IsTrue)
 	case <-time.After(coretesting.LongWait):
 		c.Fatal("timed out waiting for raft leadership")
 	}
 }
+
+func (s *RaftFixture) NewRaft(c *gc.C, id raft.ServerID, fsm raft.FSM) (
+	*raft.Raft,
+	*raft.Config,
+	*raft.InmemTransport,
+	*raft.InmemStore,
+	*raft.InmemSnapshotStore,
+) {
+	_, transport := raft.NewInmemTransport("")
+	s.AddCleanup(func(*gc.C) { transport.Close() })
+
+	store := raft.NewInmemStore()
+	snapshotStore := raft.NewInmemSnapshotStore()
+	config := s.DefaultConfig(id)
+
+	r, err := raft.NewRaft(config, fsm, store, store, snapshotStore, transport)
+	c.Assert(err, jc.ErrorIsNil)
+	s.AddCleanup(func(c *gc.C) {
+		f := r.Shutdown()
+		c.Assert(f.Error(), jc.ErrorIsNil)
+	})
+	return r, config, transport, store, snapshotStore
+}
+
+func (s *RaftFixture) DefaultConfig(id raft.ServerID) *raft.Config {
+	raftConfig := raft.DefaultConfig()
+	raftConfig.LocalID = id
+	raftConfig.HeartbeatTimeout = 100 * time.Millisecond
```

Comment:
should we set this lower for in-process testing?

Reply:
This is already lower than the default, I'm wary of lowering it more in case it starts affecting the tests on slow machines.

Reply:
I guess it isn't the send time, but more the "how long until we think someone else is dead" time. And unless we are testing the master switching, we don't really care. I wonder if it affects teardown/cleanup time. I suppose there is an explicit shutdown message.

Reply:
> I guess it isn't the send time, but more the "how long until we think someone else is dead" time. And unless we are testing the master switching, we don't really care. I wonder if it affects teardown/cleanup time. I suppose there is an explicit shutdown message.

Teardown won't be affected, but setup will. When the raft cluster is created, there's an initial election phase which apparently takes into account those timeouts. I'll drop it to 5ms (the minimum) - it does make a minor difference in the run time, and other tests can just override the config if they need something longer.

Reply:
I take that back, running with 5ms timeout under `-race` leads to test failures because of unexpected leadership changes. Given that it's only a very minor improvement, I'm going to leave it at 100ms for now. It can easily be tweaked if test runs start getting high.